atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java

```diff
@@ -76,13 +77,37 @@
      * the batch hint in each RangeRequest. If this isn't done then this method may do more work
      * than you need and will be slower than it needs to be.  If the batchHint isn't specified it
      * will default to 1 for the first page in each range.
+     *
+     * @deprecated Should use either {@link #getUnfetchedRanges(TableReference, Iterable)} or
+     * {@link #getRangesWithFirstPages(TableReference, Iterable, int, ExecutorService)} to ensure you
+     * are using an appropriate level of concurrency.
      */
     @Idempotent
+    @Deprecated
     Iterable<BatchingVisitable<RowResult<byte[]>>> getRanges(
             TableReference tableRef,
             Iterable<RangeRequest> rangeRequests);
 
     /**
+     * Returns visitibles that scan the provided ranges. This does no pre-fetching so visiting the resulting
+     * visitibles will incur database reads on first access.
+     */
+    @Idempotent
+    Iterable<BatchingVisitable<RowResult<byte[]>>> getUnfetchedRanges(
+            final TableReference tableRef, Iterable<RangeRequest> rangeRequests);
+
+    /**
+     * Same as {@link #getUnfetchedRanges(TableReference, Iterable)} but pre-fetches the first page of each visitible.
+     * @param prefetchConcurrency the maximum number of ranges that are pre-fetched concurrently
+     */
+    @Idempotent
+    Iterable<BatchingVisitable<RowResult<byte[]>>> getRangesWithFirstPages(
+            final TableReference tableRef,
+            Iterable<RangeRequest> rangeRequests,
+            int prefetchConcurrency,
+            ExecutorService executorService);
```

Comment:
I'm not in love with the user having to make their own executor service and pass it in; maybe we could just have a configurable for max number of threads for this and control a single overarching threadpool with nicely labeled threads for all of these operations in Atlas code?

Reply:
Actually @clockfort what if we did something like:
```
    public <T> Iterable<T> getRanges(
            final TableReference tableRef,
            Iterable<RangeRequest> rangeRequests,
            int concurrencyLevel,
            Function<BatchingVisitable<RowResult<byte[]>>, T> visitableProcessor)
```
so that we can pack both into one operation. Then maybe only expose that and the unfetched version. So then could do something like:
```
t.getRanges(
    table,
    rangeRequests,
    concurrencyLevel, 
    visitable -> BatchingVisitableView.of(visitable).immutableCopy());
```